targets/TARGET_NORDIC/TARGET_NRF5/trng_api.c

```diff
@@ -37,54 +37,146 @@
  */
 
 #if defined(DEVICE_TRNG)
-#include "trng_api.h"
+
+#include "hal/trng_api.h"
+#include "hal/lp_ticker_api.h"
+
 #include "nrf_drv_rng.h"
 
+#define DEFAULT_TIMEOUT_US (1000*1000)
+
+/* Macro for testing if the SoftDevice is active, regardless of whether the 
+ * application is build with the SoftDevice or not.
+ */
+#if defined(SOFTDEVICE_PRESENT)
+#include "nrf_sdm.h"
+static uint8_t wrapper(void) {
+    uint8_t softdevice_is_enabled;
+    ret_code_t result = sd_softdevice_is_enabled(&softdevice_is_enabled);
+    return ((result == NRF_SUCCESS) && (softdevice_is_enabled == 1));
+}
+#define NRF_HAL_SD_IS_ENABLED() wrapper()
+#else
+#define NRF_HAL_SD_IS_ENABLED() 0
+#endif
+
+/** Initialize the TRNG peripheral
+ *
+ * @param obj The TRNG object
+ */
 void trng_init(trng_t *obj)
 {
     (void) obj;
 
-    (void)nrf_drv_rng_init(NULL);
+    /* Initialize low power ticker. Used for timeouts. */
+    static bool first_init = true;
+
+    if (first_init) {
+        first_init = false;
+        lp_ticker_init();
+    }
 }
 
+/** Deinitialize the TRNG peripheral
+ *
+ * @param obj The TRNG object
+ */
 void trng_free(trng_t *obj)
 {
     (void) obj;
```

Comment:
@marcuschangarm Apologies if this is repeating something we already touched yesterday, but the following is still not clear to me: Why aren't we initializing the TRNG based on the SoftDevice state _at the time of the `trng_init`_ call, saving this initialization state somewhere within the `trng_t` struct, and handling `trng_get_bytes` and `trng_init` based on it (in particular, not checking whether SoftDevice has been enabled in the meantime). This way, we could correctly setup the SoftDevice / low-level TRNG drivers in `trng_init` via `nrf_drv_rng_init(NULL)` resp. `nrf_rng_task_trigger(NRF_RNG_TASK_START);`, and free them via `nrf_drv_rng_uninit` resp. `nrf_rng_task_trigger(NRF_RNG_TASK_STOP);` in `trng_free`. Is this unacceptable because we cannot guarantee that SoftDevice is enabled before `trng_init` is called, and we definitely want to use SoftDevice is possible?

Reply:
The user application can enable and disable the SoftDevice at any time. Usually this would be done to save power. On the other hand, the TRNG might be called from different parts of the application at times that are not in lock step with the SoftDevice. So the code must be able to handle init and free calls that are not in sync with the SoftDevice state.

> The SoftDevice does its own TRNG initialization. So we might end up in a situation where we use the low-level TRNG API and steal random numbers from the SoftDevice.

I believe the SoftDevice is copying the TRNG byte to a buffer on each interrupt. If we read that byte using the low level API before the SoftDevice gets to read it, weird things might happen.

Reply:
@marcuschangarm 
> I believe the SoftDevice is copying the TRNG byte to a buffer on each interrupt. If we read that byte using the low level API before the SoftDevice gets to read it, weird things might happen.
Ah, ok, so this might happen even if the SoftDevice's TRNG driver API is not used?

Reply:
> so this might happen even if the SoftDevice's TRNG driver API is not used?

Sorry, I don't understand the question.

Reply:
I understood you as saying that when SoftDevice is active, the TRNG byte is automatically read by the SoftDevice driver code when ready, even if there is no code explicitly requesting random data through `nrf_drv_rng_rand` - is that right? So if `trng_init` would be called before SoftDevice is initialized, then my suggestion for `trng_get_bytes` would lead to random data being read from the low-level driver that has potentially already been automatically processed by the SoftDevice driver code, even though `trng_api.c` didn't request that.

Reply:
> when SoftDevice is active, the TRNG byte is automatically read by the SoftDevice driver code when ready, even if there is no code explicitly requesting random data through nrf_drv_rng_rand - is that right?

Yes.

>  So if trng_init would be called before SoftDevice is initialized, then my suggestion for trng_get_bytes would lead to random data being read from the low-level driver that has potentially already been automatically processed by the SoftDevice driver code, even though trng_api.c didn't request that.

Yes, that is my concern.

Reply:
@marcuschangarm Thank you for confirming that, then I think your solution is the best one. :+1: